app/src/org/commcare/gis/MapboxLocationPickerActivity.kt

```diff
@@ -0,0 +1,206 @@
+            map.setStyle(style) {
+                val loc = viewModel.getLocation()
+                updateMarker(LatLng(loc.latitude, loc.longitude, loc.altitude))
+            }
+        }
+        current_location.setOnClickListener {
+            mapView.focusOnUserLocation(true)
+            val location = map.locationComponent.lastKnownLocation
+            location?.let {
+                updateMarker(LatLng(it.latitude, it.longitude, it.altitude))
+            }
+        }
+    }
+
+    private fun inViewMode() = intent.getBooleanExtra(GeoPointWidget.EXTRA_VIEW_ONLY, false)
+
+    override fun onResume() {
+        super.onResume()
+        observeViewModel()
+    }
+
+    private fun observeViewModel() {
+        viewModel.placeName.observe(this, androidx.lifecycle.Observer {
+            confirm_location_button.isEnabled = true
+            location.text = it
+        })
+    }
+
+    override fun onDestroy() {
+        map.removeOnMapClickListener(mapClickListener)
+        super.onDestroy()
+    }
+
+    override fun getMapLayout(): Int {
+        return R.layout.activity_mapbox_location_picker
+    }
+
+    override fun shouldFocusUserLocationOnLoad(): Boolean {
+        return false
+    }
+
+    override fun onMapLoaded() {
+        if (!inViewMode()) {
+            map.addOnMapClickListener(mapClickListener)
+        }
+        setMapStyle()
+    }
+
+    private fun setMapStyle() {
+        val style = Style.Builder()
+                .fromUri(mapStyles[currentMapStyleIndex])
+                .withImage(MARKER_ICON_IMAGE_ID, ContextCompat.getDrawable(this, R.drawable.marker)!!)
+        map.setStyle(style) {
+            loadedStyle = it
+            enableLocationComponent(it)
+            addMarker(it, map.cameraPosition.target)
+            initialMarkerPosition()
+        }
+    }
+
+    private fun enableLocationComponent(style: Style) {
+        val component = map.locationComponent
+        component.activateLocationComponent(LocationComponentActivationOptions.builder(
+                this, style
+        ).build())
+        component.isLocationComponentEnabled = true
+        component.cameraMode = CameraMode.TRACKING
+        component.renderMode = RenderMode.NORMAL
+    }
+
+    private fun addMarker(style: Style, point: LatLng) {
+        symbolManager = SymbolManager(mapView, map, style)
+        symbolManager?.let { symbolManager ->
+            symbolManager.iconAllowOverlap = true
+            val symbolOptions = SymbolOptions()
+                    .withLatLng(point)
+                    .withIconImage(MARKER_ICON_IMAGE_ID)
+            symbol = symbolManager.create(symbolOptions)
+        }
+    }
+
+    private fun initialMarkerPosition() {
+        if (cameraPosition != null) {
+            cameraPosition?.let { pos ->
+                viewModel.reverseGeocode(pos.target)
+                updateMarker(pos.target)
+            }
+        } else {
+            map.locationComponent.locationEngine
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/dimagi/commcare-android/blob/cb307f0c79a9e28d4bce7b60e4ee756470e4a32e/app/src/org/commcare/gis/MapboxLocationPickerActivity.kt#L179"></span>Call requires permission which may be rejected by user: code should explicitly check to see if permission is available (with <code>checkPermission</code>) or explicitly handle a potential <code>SecurityException</code></td>
    </tr>
  </tbody>
</table>